sonar-dotnet-shared-library/src/main/java/org/sonar/plugins/dotnet/tests/ScannerFileService.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * SonarSource :: .NET :: Shared library
+ * Copyright (C) 2014-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.dotnet.tests;
+
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.sonar.api.batch.fs.FilePredicates;
+import org.sonar.api.batch.fs.FileSystem;
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.scanner.ScannerSide;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+@ScannerSide
+public class ScannerFileService implements FileService {
+  private static final Logger LOG = Loggers.get(ScannerFileService.class);
+  private static final Pattern DETERMINISTIC_SOURCE_PATH_PREFIX = Pattern.compile("^(/_\\d*/)");
+  private FileSystem fileSystem;
+  private String languageKey;
+
+  public ScannerFileService(String languageKey, FileSystem fileSystem) {
+    this.languageKey = languageKey;
+    this.fileSystem = fileSystem;
+  }
+
+  public boolean isSupportedAbsolute(String absolutePath) {
+    FilePredicates fp = fileSystem.predicates();
+    return fileSystem.hasFiles(
+      fp.and(
+        fp.hasAbsolutePath(absolutePath),
+        fp.hasLanguage(languageKey)));
+  }
+
+  public Optional<InputFile> getFileByRelativePath(String filePath) {
+    Matcher matcher = DETERMINISTIC_SOURCE_PATH_PREFIX.matcher(filePath.replace('\\', '/'));
+    if (matcher.find()) {
+      String relativePath = matcher.replaceFirst("");
+      Iterable<InputFile> files = fileSystem.inputFiles(fileSystem.predicates().hasLanguage(languageKey));
+      int count = 0;
+      InputFile foundFile = null;
+      for (InputFile file : files) {
+        String path = file.uri().getPath();
+        if (path.endsWith(relativePath)) {
+          count++;
+          foundFile = file;
+        }
+      }
```

Comment:
This new layout is great, as it's easy to follow. Now it's way more visible what's going.

_Open question:_ The filtering in the `for` is actually a manual way of implementing lambda. `predicate` actually in Java terms. So can we replace this section with something predicate based?

There exists `predicates().hasRelativePath()`. Does that one solve our issue (I'm not sure if it does the same)? If not, isn't it simpler to implement a predicate? Even in-lined one with the `new FilePredicate(){accept(...)}` java syntax for inline abstract class implementation. As it should be 3 LOC.

We would not need the `count` variable.
We would return to the structure I asked you to remove previously :D 
```
    FilePredicates fp = fileSystem.predicates();
    Iterable<InputFile> files = fileSystem.inputFiles(
      fp.and(
        fp.hasRelativePath(path),
        fp.hasLanguage(languageKey)
      ));
```

Reply:
as previously mentioned during a standup on Friday, this doesn't work. This is why I ended up doing this logic.

Reply:
I read your suggestion better now, getting it back in progress

Reply:
Copy from Slack, as I just found out:
The relative path predicate does this:
```
return path.equals(f.relativePath()); 
```
so it will not help us :disappointed:
As seen here: https://github.com/SonarSource/sonar-enterprise/blob/021961b86da874e4b733a566ce252[…]nar/api/batch/fs/internal/predicates/RelativePathPredicate.java
PathPatternPredicate could do the trick, but it doesn't worth the trouble with crafting the dynamic regex
So if you see easy way how to implement own predicate, let's do it. Otherwise, let's merge it as is.

Reply:
PathPatternPredicate = predicates().matchesPathPattern()